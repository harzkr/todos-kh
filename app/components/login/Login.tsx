"use client";
import React, { useState } from "react";
import { Typography, TextField, Button, CircularProgress } from "@mui/material";
import cssStyles from "./login.module.css";
import { useUserLoginMutation } from "@/lib/features/login/loginApiSlice";
import { useRouter } from "next/navigation";

export const Login = () => {
  const router = useRouter();

  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const [loggingIn, setLoggingIn] = useState(false);

  const [loginUser, { data, error, isLoading }] = useUserLoginMutation();

  const handleLogin = async () => {
    setErrorMessage("");
    setLoggingIn(true);
    await loginUser({ username, password });
  };

  const handleRedirection = async () => {
    setLoggingIn(false);
    router.replace("/");
  };

  React.useEffect(() => {
    if (data && data.string) {
      console.log("data", data);
      if (data.string === "ok") {
        // redirect to dashboard
        handleRedirection();
      } else {
        // show error message
        console.log("error", data);
        if (data.error_message) {
          setErrorMessage(data.error_message);
          setLoggingIn(false);
        }
      }
    }

    if (error) {
      console.log("error", error);
      setErrorMessage("An error occurred. Please try again.");
      setLoggingIn(false);
    }
  }, [data, error]);

  if (loggingIn) {
    return <CircularProgress />;
  }

  return (
    <div className={cssStyles.loginContainer}>
      <Typography variant="body2">ENTER YOUR CREDENTIALS TO LOGIN</Typography>
      {errorMessage && (
        <Typography variant="body2" color="error">
          {errorMessage}
        </Typography>
      )}
      <div className={cssStyles.outerContainer}>
        <TextField
          id="username"
          label="Username"
          variant="outlined"
          onChange={(e) => setUsername(e.target.value)}
          color="secondary"
          fullWidth
          data-testid="username"
        />
        <br />
        <TextField
          id="password"
          label="Password"
          type="password"
          variant="outlined"
          onChange={(e) => setPassword(e.target.value)}
          color="secondary"
          fullWidth
          data-testid="password"
        />
      </div>
      <Button
        variant="text"
        disabled={username.length === 0 || password.length === 0 || isLoading}
        onClick={handleLogin}
        color="secondary"
        data-testid="login-button"
      >
        Login
      </Button>
    </div>
  );
};
