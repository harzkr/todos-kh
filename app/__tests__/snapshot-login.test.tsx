import "whatwg-fetch";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import Login from "../login/page";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
}));

const mockStore = configureStore();

describe("Login", () => {
  it("renders homepage unchanged", () => {
    const { container } = render(
      <Provider store={mockStore()}>
        <Login />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
