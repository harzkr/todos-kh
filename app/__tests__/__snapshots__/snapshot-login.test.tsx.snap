// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders homepage unchanged 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    style="background-image: url(/img.jpg); height: 100vh; width: 100%; background-size: cover; background-position: center; background-repeat: no-repeat; display: flex; justify-content: center; align-items: center; flex-direction: column;"
  >
    <div
      class="loginContainer"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
      >
        ENTER YOUR CREDENTIALS TO LOGIN
      </p>
      <div
        class="outerContainer"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="username"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1v4qvbo-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="username"
            id="username-label"
          >
            Username
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-fullWidth MuiInputBase-formControl css-1j0w0ul-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="username"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Username
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <br />
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="password"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1v4qvbo-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiInputBase-fullWidth MuiInputBase-formControl css-1j0w0ul-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary Mui-disabled MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary css-1d3f8j8-MuiButtonBase-root-MuiButton-root"
        data-testid="login-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
